code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -298,4 +312,51 @@ public boolean isGlobalService(Service service) {
         }
         return false;
     }
+
+   @Override
+   public void updateServicePublicEndpoints(final Service service, final IpPortPublicEndpoint publicEndpoint, final boolean add) {
+        lockManager.lock(new ServiceEndpointsUpdateLock(service), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                updateObjectEndPoint(service, service.getKind(), service.getId(), publicEndpoint, add);
+            }
+        });
+    }
+
+    @Override
+    public void updateHostPublicEndpoints(final Host host, final IpPortPublicEndpoint publicEndpoint, final boolean add) {
+        lockManager.lock(new HostEndpointsUpdateLock(host), new LockCallbackNoReturn() {
+            @Override
+            public void doWithLockNoResult() {
+                updateObjectEndPoint(host, host.getKind(), host.getId(), publicEndpoint, add);
+            }
+        });
+    }
+
+    protected void updateObjectEndPoint(final Object service, String resourceType, Long resourceId,
+            final IpPortPublicEndpoint publicEndpoint, final boolean add) {
+        List<IpPortPublicEndpoint> publicEndpoints = new ArrayList<>();
+        publicEndpoints.addAll(DataAccessor.fields(service)
+                .withKey(ServiceDiscoveryConstants.FIELD_PUBLIC_ENDPOINTS)
+                .withDefault(Collections.EMPTY_LIST)
+                .asList(jsonMapper, IpPortPublicEndpoint.class));
+        if (publicEndpoints.contains(publicEndpoint) != add) {
+            if (add) {
+                publicEndpoints.add(publicEndpoint);
+            } else {
+                publicEndpoints.remove(publicEndpoint);
+            }
+            DataUtils.getWritableFields(service).put(ServiceDiscoveryConstants.FIELD_PUBLIC_ENDPOINTS,
+                    publicEndpoints);
+            objectManager.persist(service);
+            Map<String, Object> data = new HashMap<>();
+            data.put(ServiceDiscoveryConstants.FIELD_PUBLIC_ENDPOINTS, publicEndpoints);
```

Comment:
We've copied and pasted this code in a couple places now.  Can we make it generic somewhere.  So you can just call `utils.TriggerStateChanged(eventService, resource)`.  And the call to eventService.publish should be deferred.